### PR TITLE
#6 Task. Yura Matsiulka

### PR DESCRIPTION
**Pull request by Happy-Love**
https://travis-ci.org/github/Happy-Love/js-assignments/builds/709994594

### CODE REVIEW

File: task/06-conditions-n-loops-tasks.js [hunk 1]
```diff
@@ -45,8 +48,20 @@ function getFizzBuzz(num) {
  *   5  => 120
  *   10 => 3628800
  */
+
 function getFactorial(n) {
-    throw new Error('Not implemented');
+    
+    function factorialize(num) {
```
Comment (AlexMokin) [inline task/06-conditions-n-loops-tasks.js]: there is no point of creating another functions here

File: task/06-conditions-n-loops-tasks.js [hunk 2]
```diff
@@ -270,7 +315,14 @@ function isCreditCardNumber(ccn) {
  *   165536 (1+6+5+5+3+6 = 26,  2+6 = 8) => 8
  */
 function getDigitalRoot(num) {
```
Comment (AlexMokin) [inline task/06-conditions-n-loops-tasks.js]: fix indentation in this method

File: task/06-conditions-n-loops-tasks.js [hunk 3]
```diff
@@ -431,7 +533,46 @@ function getMatrixProduct(m1, m2) {
...
+           let truth = 0;
+           for (let j = 0; j < n; j++) {
+                for (let k = 0; k < arr.length; k++ ) {
+                    if ( winCoords[i][j] === arr[k])
+                        truth += 1    
+                }
+           }
+           if (truth === 3) return symb; 
+       }
+    return false;
```
Comment (AlexMokin) [inline task/06-conditions-n-loops-tasks.js]: indentation

File: task/06-conditions-n-loops-tasks.js [hunk 4]
```diff
@@ -431,7 +533,46 @@ function getMatrixProduct(m1, m2) {
...
+                if (position[i-1][j-1] === symb)
+                    arr.push(Number(i.toString() + j));
+            }
+        }
+        return arr;   
+    }
+
+    function isWin(arr, symb) {
+       for (let i = 0; i < winCoords.length; i++) {
+           let truth = 0;
```
Comment (AlexMokin) [inline task/06-conditions-n-loops-tasks.js]: bad variable name

File: task/06-conditions-n-loops-tasks.js [hunk 5]
```diff
@@ -431,7 +533,46 @@ function getMatrixProduct(m1, m2) {
...
+        [11,22,33], [13,22,31]
+    ];
+    let n = 3;
+
+    function transformMatrix(position, symb) {
+        let arr = [];
+        for(let i = 1; i <= n; i++){
+            for(let j = 1; j <= n; j++) {
+                winCoords[i][j] === arr[j];
+                if (position[i-1][j-1] === symb)
```
Comment (AlexMokin) [inline task/06-conditions-n-loops-tasks.js]: just use `for(let i = 0; i < n; i++)` and won't need to subtract 1

File: task/06-conditions-n-loops-tasks.js [hunk 6]
```diff
@@ -431,7 +533,46 @@ function getMatrixProduct(m1, m2) {
...
+        [11,21,31], [12,22,32], [13,23,33],
+        [11,22,33], [13,22,31]
+    ];
+    let n = 3;
+
+    function transformMatrix(position, symb) {
+        let arr = [];
+        for(let i = 1; i <= n; i++){
+            for(let j = 1; j <= n; j++) {
+                winCoords[i][j] === arr[j];
```
Comment (AlexMokin) [inline task/06-conditions-n-loops-tasks.js]: that line doesn't do anything

File: task/06-conditions-n-loops-tasks.js [hunk 7]
```diff
@@ -431,7 +533,46 @@ function getMatrixProduct(m1, m2) {
...
+                    if ( winCoords[i][j] === arr[k])
+                        truth += 1    
+                }
+           }
+           if (truth === 3) return symb; 
+       }
+    return false;    
+    }
+
+    let arr = transformMatrix(position, 'X');
```
Comment (AlexMokin) [inline task/06-conditions-n-loops-tasks.js]: you don't really care if it's an X or O . Just go over the wining positions and make sure they have the same value and just  get one this values.

**Review by AlexMokin — COMMENTED** [review]
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

File: task/06-conditions-n-loops-tasks.js [hunk 8]
```diff
@@ -45,8 +48,15 @@ function getFizzBuzz(num) {
...
  *   10 => 3628800
  */
+
 function getFactorial(n) {
-    throw new Error('Not implemented');
+    // Status OK (1)
+    if (n === 0 || n === 1)
+      return 1;
+    for (let i = n - 1; i >= 1; i--) {
+       n *= i;
```
Comment (AlexMokin) [inline task/06-conditions-n-loops-tasks.js]: do not modify arguments if not requred

File: task/06-conditions-n-loops-tasks.js [hunk 9]
```diff
@@ -269,8 +309,16 @@ function isCreditCardNumber(ccn) {
  *   10000 ( 1+0+0+0+0 = 1 ) => 1
  *   165536 (1+6+5+5+3+6 = 26,  2+6 = 8) => 8
  */
-function getDigitalRoot(num) {
-    throw new Error('Not implemented');
+function getDigitalRoot(num){
+    //(fixed)
+    if (num < 9) 
+        return num;
```
Comment (AlexMokin) [inline task/06-conditions-n-loops-tasks.js:315]: brackets

File: task/06-conditions-n-loops-tasks.js [hunk 10]
```diff
@@ -296,7 +344,20 @@ function getDigitalRoot(num) {
...
 function isBracketsBalanced(str) {
-    throw new Error('Not implemented');
+    let arr = [];
+    let open = ['[', '(', '<', '{'];
+    let close = [']', ')', '>', '}'];
+
+    for (let i = 0; i < str.length; i++) {     
+        if (open.indexOf(str[i]) !== -1) {
+            arr.push(str[i]);
+            continue
```
Comment (AlexMokin) [inline task/06-conditions-n-loops-tasks.js]: semicolon

File: task/06-conditions-n-loops-tasks.js [hunk 11]
```diff
@@ -296,7 +344,20 @@ function getDigitalRoot(num) {
...
+        }
+        
+        if (open.indexOf(arr.pop()) !== close.indexOf(str[i]))  return false;
```
Comment (AlexMokin) [inline task/06-conditions-n-loops-tasks.js]: too long, move return false on the nesxt line, and add brackets

File: task/06-conditions-n-loops-tasks.js [hunk 12]
```diff
@@ -431,7 +529,31 @@ function getMatrixProduct(m1, m2) {
...
+        [11,12,13], [21,22,23], [31,32,33],
+        [11,21,31], [12,22,32], [13,23,33],
+        [11,22,33], [13,22,31]
+    ];
+    // Привет Алекс, не универсально но грубо говоря я скоротил решение на половину узнав выигрышные координаты заранее.
+    // Какая логика?
+    // Проверяем заранее известные координаты победы.
+    // 11 12 13
+    // Логика проверки
+    // Каждый раз нам придётся сверять по 3 отметки, если хотя бы 1 из комбинации является выигрышной значит это победа одного из маркеров
```
Comment (AlexMokin) [inline task/06-conditions-n-loops-tasks.js]: english

File: task/06-conditions-n-loops-tasks.js [hunk 13]
```diff
@@ -431,7 +529,31 @@ function getMatrixProduct(m1, m2) {
...
+    let n = 3;
+    let firstLetter="";
+    for(let i=0;i<winCoords.length;i++){
+        let mark=0;
+        firstLetter=position[Number(winCoords[i][0].toString()[0])-1][Number(winCoords[i][0].toString()[1])-1];
```
Comment (AlexMokin) [inline task/06-conditions-n-loops-tasks.js]: this is not very readable why don't just code win positions as
[[1,1],[1,2],[1,3]]
?

or just write a function that gets element form position array

File: task/06-conditions-n-loops-tasks.js [hunk 14]
```diff
@@ -431,7 +529,31 @@ function getMatrixProduct(m1, m2) {
...
+        if(firstLetter === undefined || firstLetter===null || firstLetter==='' || firstLetter===' ') continue;
+        for (let j = 0; j < winCoords[i].length; j++) {
+            if(position[Number(winCoords[i][j].toString()[0])-1][Number(winCoords[i][j].toString()[1])-1]===firstLetter)
+               mark++;   
+        }
+        if(mark===3)
+            return firstLetter;
```
Comment (AlexMokin) [inline task/06-conditions-n-loops-tasks.js:570]: brackets

**Review by AlexMokin — COMMENTED** [review]
Review this https://javascript.info/coding-style and fix your code accoringly to this code style.
- make sure you use 4 spaces for indetntations
- check for missing semicolons
- check for missing brackets

Also see comments below

I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

File: task/06-conditions-n-loops-tasks.js (hunk 13 above)
Comment (Happy-Love) [reply task/06-conditions-n-loops-tasks.js]: because this is my code))

File: task/06-conditions-n-loops-tasks.js (hunk 11 above)
Comment (Happy-Love) [reply task/06-conditions-n-loops-tasks.js]: Understand

File: task/06-conditions-n-loops-tasks.js (hunk 10 above)
Comment (Happy-Love) [reply task/06-conditions-n-loops-tasks.js]: Very strong error

File: task/06-conditions-n-loops-tasks.js (hunk 9 above)
Comment (Happy-Love) [reply task/06-conditions-n-loops-tasks.js:315]: Understand

File: task/06-conditions-n-loops-tasks.js [hunk 15]
```diff
@@ -192,7 +210,12 @@ function findFirstSingleChar(str) {
  *
  */
 function getIntervalString(a, b, isStartIncluded, isEndIncluded) {
-    throw new Error('Not implemented');
+    let out = '';
+    out += (isStartIncluded)? '[' : '(' ;
+    if (a < b) out +=  a + ', ' + b ;
```
Comment (AlexMokin) [inline task/06-conditions-n-loops-tasks.js:215]: if you do an assignment it's better to move the body on the next line and use brackets

File: task/06-conditions-n-loops-tasks.js (hunk 13 above)
Comment (AlexMokin) [reply task/06-conditions-n-loops-tasks.js]: yeap, but keep in mind that somebody's going to read it

**Comment by AlexMokin** [comment]
Accepted, but 
- go over the article one more time. https://javascript.info/coding-style
   Emphasis on spaces around operators.
- do not try to save lines with one-line ifs, for loops etc.

**Pull request closed without merging**